test/langtools/tools/javac/reflect/MethodReferenceTest.java

```diff
@@ -182,13 +182,13 @@ void test7() {
     @CodeReflection
     @IR("""
             func @"test8" (%0 : MethodReferenceTest)void -> {
-                %1 : java.util.function.IntFunction<MethodReferenceTest$A<java.lang.String>[]> = lambda (%2 : int)MethodReferenceTest$A<java.lang.String>[] -> {
+                %1 : java.util.function.IntFunction<.<MethodReferenceTest, MethodReferenceTest$A<java.lang.String>>> = lambda (%2 : int).<MethodReferenceTest, MethodReferenceTest$A<java.lang.String>> -> {
```

Comment:
Missing `[]`?

```suggestion
                %1 : java.util.function.IntFunction<.<MethodReferenceTest, MethodReferenceTest$A<java.lang.String>[]>> = lambda (%2 : int).<MethodReferenceTest, MethodReferenceTest$A<java.lang.String>[]> -> {
```

Reply:
uhmmm - the test is passing but you are correct that there should be some brackets in here. I'll investigate

Reply:
Hmm... i suspect the it's due to canonicalization. We parse the expected model and serialize it out. Depending on the operation its result type is determined by say its operands rather than what is explicitly declared in the text.

Reply:
> Hmm... i suspect the it's due to canonicalization. We parse the expected model and serialize it out. Depending on the operation its result type is determined by say its operands rather than what is explicitly declared in the text.

That might explain why I cannot reproduce compiling the test and dumping out the IR...

Reply:
> > Hmm... i suspect the it's due to canonicalization. We parse the expected model and serialize it out. Depending on the operation its result type is determined by say its operands rather than what is explicitly declared in the text.
> 
> That might explain why I cannot reproduce compiling the test and dumping out the IR... but it used to work fine, so I doubt it's the op itself. There must be something flaky in the serialization/deserialization process that's leaving `[]` behind.

This kind of give me an idea to add a test that does exactly this: takes a type, serialize it, deserialize it back, and check it's the same thing.

Reply:
Good idea!

Reply:
Defo seeing some issues with the new test:
```
expected [java.util.List<#TestJavaType$TypeHolder::X<java.lang.Number>>] but found [java.util.List<#TestJavaType$TypeHolder::X<java.lang.Number>[]>]
```

Reply:
The JavaType factory is now fixed. This turned into a biggie overhaul of the factory code.
Separately (e.g. in a different PR), I think it's time we look again at the `toString` representation of JavaType. Right now we return the external form, but I think we can do better, and eliminate a lot of visual clutter. What do you think?